TruthTables/Tasks.qs

```diff
@@ -0,0 +1,124 @@
+    //          and 111.
+    operation ProjectiveTruthTables () : (TruthTable, TruthTable, TruthTable) {
+        let x1 = TruthTable(0b10101010, 3);
+        let x2 = TruthTable(0, 0);           // Update the value of x2 ...
+        let x3 = TruthTable(0, 0);           // Update the value of x3 ...
+
+        return (x1, x2, x3);
+    }
+
+    // Task 2. Compute AND of two truth tables
+    //
+    // Goal: Compute a truth table that computes the conjunction (AND)
+    //       of two truth tables.
+    //
+    // Hint: You can use bit-wise operations in Q# for this task.  For
+    //       AND the fitting operation is `&&&`.
+    operation TTAnd (tt1 : TruthTable, tt2 : TruthTable) : TruthTable {
+        let (bits1, numVars1) = tt1!;
+        let (bits2, numVars2) = tt2!;
+        EqualityFactI(numVars1, numVars2, "Number of variables for both truth tables must match");
+
+        fail ("Task 2 not implemented!");
+    }
+
+    // Task 3. Compute OR of two truth tables
+    //
+    // Goal: Compute a truth table that computes the disjunction (OR)
+    //       of two truth tables.
+    operation TTOr (tt1 : TruthTable, tt2 : TruthTable) : TruthTable {
+        fail ("Task 3 not implemented!");
+    }
+
+    // Task 4. Compute XOR of two truth tables
+    //
+    // Goal: Compute a truth table that computes the exclusive-OR (XOR)
+    //       of two truth tables.
+    operation TTXor (tt1 : TruthTable, tt2 : TruthTable) : TruthTable {
+        fail ("Task 4 not implemented!");
+    }
+
+    // Task 5. Compute NOT of a truth table
+    //
+    // Goal: Compute a truth table that computes negation of a truth
+    //       table.
+    //
+    // Hint: Be careful not to set bits in the integer that are out-of-range
+    //       in the truth table.
+    operation TTNot (tt : TruthTable) : TruthTable {
+        fail ("Task 5 not implemented!");
+    }
+
+    // Task 6. Build if-then-else truth table
+    //
+    // Goal: Compute the truth table of the if-then-else function x1 ? x2 : x3
+    //       (if x1 then x2 else x3) by making use of the truth table operations
+    //       defined in the previous 4 tasks.
+    operation TTIfThenElse (ttCond : TruthTable, ttThen : TruthTable, ttElse : TruthTable) : TruthTable {
+        fail ("Task 6 not implemented!");
+    }
+
+    // Task 7. Find all true input assignments in a truth table
+    //
+    // Goal: Return an array that contains all input assignments for
+    //       function represented as truth table.  Make use of Q# library functions
+    //       to implement this operation without implementing any helper operations.
+    //       Useful Q# library functions to complete this task are Mapped, Filtered,
+    //       Compose, Enumerated, IntAsBoolArray, EqualB, Fst, and Snd.
+    //
+    // Example: The truth table of 2-input OR is 0b1110, i.e., its minterms are
+    //          [1, 2, 3].
+    operation AllMinterms (tt : TruthTable) : Int[] {
```

Comment:
I would suggest making this task (and the previous ones) functions instead of operations - this allows to solve the last task as 

    for (i in AllMinterms(tt)) {
        (ControlledOnInt(i, X))(controls, target);
    }

with adjoint operation autogenerated (if it is an operation, automatic generation fails). And then you won't need the hint about the operation being self-inverse, since is Adj will take care of it.

Reply:
Good point. I was wondering why automatic generation failed.